File: 0463-island-perimeter/0463-island-perimeter.py
class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        
        m, n = len(grid), len(grid[0])
        ans = 0
        for x in range(m):
            for y in range(n):
                for dx, dy in [(0, -1), (-1, 0)]:
                    nx, ny = x + dx, y + dy
                    if(0 <= nx < m and 0 <= ny < n):
                        ans += grid[x][y] != grid[nx][ny]
                    else:
                        ans += grid[x][y]
                        
        for x in range(n):
            ans += grid[m - 1][x]
            
        for y in range(m):
            ans += grid[y][n - 1]
                        
        return ans
        

File: 0463-island-perimeter/0463-island-perimeter.go
func islandPerimeter(grid [][]int) int {
    rows := len(grid)
    cols := len(grid[0])
	
	set := make(map[[2]int]bool)
    
    for r := 0; r < rows; r++ {
        for c := 0; c < cols; c++{
            if grid[r][c] == 1 {
                return findPerimeter(grid, r, c, set)
            }
        }
    }
    return 0
}

func findPerimeter(grid [][]int, sr, sc int, set map[[2]int]bool) int{
    rows := len(grid)
    cols := len(grid[0])
	coor := [2]int {sr, sc}

	if set[coor] {
		return 0
	}

    if (sr < 0 || sr >= rows || 
        sc < 0 || sc >= cols || 
        grid[sr][sc] == 0){   
        return 1
    }

	set[coor] = true
    dire := [5]int {0, 1, 0, -1, 0}
    var ret int;
    for i := range dire[:len(dire) - 1]{
		
        ret += findPerimeter(grid, sr + dire[i], sc + dire[i + 1], set)
    }
    return ret
}